### DISCUSSION TITLE
Ability to have custom field permission based on item value

### DISCUSSION THREAD

**Discussion by 8byr0** [directus/directus#9344] [Archive]
Hey there!

I'm working on a project with an `invoice` collection like so:
```typescript
interface Invoice {
    locked: Boolean;
    payment_date: Boolean;
    amount: Number;
    // Other fields
}
```

My workflow is simple: whenever the invoice is locked, users are only allowed to update `payment_date` field.
In a custom role, I have the following rule under `edit` > `item permissions`
```json
{
	"locked": {
		"_eq": false
	}
}
```
![image](https://user-images.githubusercontent.com/21955112/139832901-b59df99e-20bb-48ac-b361-79799e96be64.png)

All fields are selected under `field permissions` tab.

With this configuration, whenever an invoice is updated with `locked` set to `true` updating any field is no longer possible which is cool BUT I'd like to allow `payment_date` field to be still updatable even after lock.

From what I've read in docs, discussions and issues, I could not find any relevant example for my use case.

How could I manage such behavior? Is there an easy way to do it or should I make a tricky extension?

**Comment by 8byr0** [comment]
I use to have 2 hobbies in my life: coding and answering my own questions 😄 

Here's how I figured out how to handle this if anyone reaches this topic with a similar need:
1. Update permissions to only include ownership behavior (e.g. only creator can CRUD)
2. write a simple hook that will cleanup payload before update, depending on actual data
```javascript
// extensions/hooks/lock-invoice.js

module.exports = function registerHook({ services }) {
  const { ItemsService } = services;

  return {
    "items.update.before": async function (
      input,
      { collection, item, schema, accountability }
    ) {
      if (collection !== "invoice") return input;
      // If an admin we allow update
      if (accountability.admin === true) return input;

      // Checkout invoice data
      const invoiceService = new ItemsService("invoice", { schema });
      const existingInvoice = await invoiceService.readOne(item[0]);

      // If locked, drop all keys not allowed
      if (existingInvoice.locked === true) {
        const allowedKeys = ["payment_date"];

        Object.keys(input).forEach((inputKey) => {
          if (!allowedKeys.includes(inputKey)) {
            console.log("Deleting forbidden key", inputKey);
            delete input[inputKey];
          }
        });
      }

      return input;
    },
  };
};
```



This hook surely does the job but is not optimal since it kind of lies to the client: a success response will be returned without additional notice. I decided to do it this way for a soft behavior but another approach could be to throw an error when a forbidden key is found, instead of silently ignoring it:
```javascript
// extensions/hooks/lock-invoice.js

module.exports = function registerHook({ services, exceptions }) {
  const { ItemsService } = services;
  const { InvalidPayloadException } = exceptions;

  return {
    "items.update.before": async function (
      input,
      { collection, item, schema, accountability }
    ) {
      if (collection !== "invoice") return input;
      // If an admin we allow update
      if (accountability.admin === true) return input;

      // Checkout invoice data
      const invoiceService = new ItemsService("invoice", { schema });
      const existingInvoice = await invoiceService.readOne(item[0]);

      // If locked, store all forbidden keys in order to throw a verbose error
      if (existingInvoice.locked === true) {
        const allowedKeys = ["payment_date"];
        const foundForbiddenKeys = [];
        Object.keys(input).forEach((inputKey) => {
          if (!allowedKeys.includes(inputKey)) {
            console.log("Found a forbidden key", inputKey);
            foundForbiddenKeys.push(inputKey);
          }
        });
        if (foundForbiddenKeys.length > 0) {
          throw new InvalidPayloadException(
            `Your payload contains invalid data: ${foundForbiddenKeys} - allowed keys: ${allowedKeys}`
          );
        }
      }

      return input;
    },
  };
};
```

**Comment by rijkvanzanten** [comment]
Heya! Good one, this is currently not possible. Here's another discussion on the topic (in less detail though) https://github.com/directus/directus/discussions/6099.

Permissions control whether or not the user is allowed to update an item, and the fields select what fields the user is allowed to update when they have permissions to update the item. However, you can't currently control what fields the user is allowed to update based on different permission rules. You could work around it by using conditional rules on every field to check "If locked = true, set readonly = true" though that would only be on the presentation layer in the app, so you'd still need the custom hook above to do the API side validation 👍🏻

**Reply by 8byr0** [reply]
The project this hook is for will actually not be using directus dashboard at all (except for admin purpose on my side, with all permissions) and implementing security on client side is actually something I would never rely on 🔒 

I think this issue may be addressed differently when each user will be able to have multiple roles, to be continued !

**Reply by rijkvanzanten** [reply]
> and implementing security on client side is actually something I would never rely on 🔒

Oh yeah definitely, not suggesting you'd rely on that exclusively! Just mentioning it as a way to optimize the admin editing experience for your users while relying on the custom hook for the actual validation 🙂 That being said, seeing you won't have "regular" users interface with the admin app, feel free to ignore that suggestion.
